tests/types/PropsWithType.test.ts

```diff
@@ -0,0 +1,28 @@
+import { expect } from 'chai';
+import { PropsWithType } from '../../src/index';
+
+describe('PropsWithType', () => {
+
+   // NOTE This test does not really check much at runtime. The real value is in the type
+   // checking that happens by *using* the `PropsWithType` type in this test. If we break
+   // the definition of `PropsWithType` so that it no longer selects the correct
+   // properties, the TypeScript compiler will complain about this test.
+   it('allows property keys that have a given type', () => {
+      interface TestType {
+         readonly prop: string;
+         readonly anotherProp: number;
+         readonly additionalProp: number;
+      }
+
+      function getNumber(from: TestType, field: PropsWithType<TestType, number>): number {
+         // `field` will only be allowed to be `anotherProp` or `additionalProp`
+         return from[field];
+      }
+
+      let obj: TestType = { prop: 'test', anotherProp: 1, additionalProp: 2 };
+
+      expect(getNumber(obj, 'anotherProp')).to.strictlyEqual(1);
```

Comment:
ADIRL: it'd be nice to actually validate that `getNumber(obj, 'prop')` and `getNumber(obj, 'doesNotExist')` both fail.

Reply:
I think something like [`dtslint`](https://github.com/microsoft/dtslint) has promise for this type of testing. I've pushed a [WIP example here](https://github.com/silvermine/toolbox/commit/1d627c8a8a1973acaa7649b27396e533784921d2). However, as that will be more work to implement, I vote we do that work in a later PR.